README.md
```
                 __ __   __                                    
.--------.--.--.|  |  |_|__|.-----.---.-.----.-----.-----.----.
|        |  |  ||  |   _|  ||  _  |  _  |   _|__ --|  -__|   _|
|__|__|__|_____||__|____|__||   __|___._|__| |_____|_____|__|  
                            |__|
```
A streams2 compatible multipart-form parser. Hacked from [formidable](https://github.com/felixge/node-formidable.git).

## Why
Formidable currently doesn't work with node >= 9.4. Also a good excuse to learn streams2.

## How
The module exports a class `Multiparser` which inherits from `Stream.Writable`. Http request streams can be piped to instances which will emit `'part'` events (essentially instances of `Stream.PassThrough`) that can then be piped around as needed (to file, s3, etc). Backpressure from parts' destinations are magically transmitted back to the Multiparser's source.

## Install
`npm install multiparser`

## Usage
The code below is pretty much straight from `example/http.js`
```javascript
var fs = require('fs')
var Multiparser = require('multiparser')
var http = require("http")
var updir = '/tmp/'

http.createServer(function (req, res) {
  if (req.url === '/upload' && req.method === 'POST') {

    var error = null
    var progress = 0
    var parser = new Multiparser(req)

    parser.on('error', function (err) {
      error = err
    })

    parser.on('progress', function (parsed, total) {
      var currentProgress = ~~(parsed / total * 100)
      if (currentProgress != progress) {
        console.log('uploading...', currentProgress, req._readableState.length)
        progress = currentProgress
      }
    })

    parser.on('part', function (part) {
      part.on('end', function () {
        if (part.filename) {
          console.log('uploaded file "' + part.name + '" to ' + updir + part.filename)
        } else if (part.value) {
          console.log('parsed field "' + part.name + '" as "' + part.value + '"')
        }
      })
      
      if (part.filename) {
        var file = fs.createWriteStream(updir + part.filename)
        part.pipe(file)
      } else {
        part.value = ''
        part.on('readable', function () {
          part.value += part.read()
        })
        part.read(0)
      }
    })

    parser.on('end', function () {
      if (error) {
        res.writeHeaders(500)
        return res.end(err.message)
      } else {
        res.end('all parsed!')
      }
    })

  } else {

    res.end('<html>\
      <form enctype="multipart/form-data" method="POST" action="/upload">\
      <input name="fieldup" type="text" /><br>\
      <input name="fileup" type="file" multiple="true" /><br>\
      <input type="submit" value="upload" />\
      </form>\
      </html>')

  }
}).listen(8080)
```

## License
MIT

HISTORY.md
## 0.0.5 / 2013 1 9

* fixed a bug in lib/Transformer which could cause the parser to stop midstream
* more refactoring

## 0.0.4 / 2013 1 9

* refactored into three classes - Multiparser, Part & Transformer

## 0.0.1 / 2013 1 1

* first release

lib/multiparser.js
/*
 *  multiparser.js
 *
 */

// export Multiparser class
module.exports = Multiparser

var util = require('util'),
    stream = require('stream'),
    Part = require('./part'),
    s = 0,
    S = { 
      PARSER_UNINITIALIZED: s++,
      START: s++,
      START_BOUNDARY: s++,
      HEADER_FIELD_START: s++,
      HEADER_FIELD: s++,
      HEADER_VALUE_START: s++,
      HEADER_VALUE: s++,
      HEADER_VALUE_ALMOST_DONE: s++,
      HEADERS_ALMOST_DONE: s++,
      PART_DATA_START: s++,
      PART_DATA: s++,
      PART_END: s++,
      END: s++
    },
    f = 1,
    F = { 
      PART_BOUNDARY: f,
      LAST_BOUNDARY: f *= 2
    },
    LF = 10,
    CR = 13,
    SPACE = 32,
    HYPHEN = 45,
    COLON = 58,
    A = 97,
    Z = 122

for (s in S) {
  exports[s] = S[s]
}

function lower(c) {
  return c | 0x20
}

function Multiparser(request, opts) {
  var self = this
  stream.Writable.call(this, opts)
  
  this.parseHeaders(request.headers)
  if (this.type !== 'multipart') {
    this.emit('error', new Error('content-type is not \'multipart/form-data\''))
    return this.end()
  } else if (!this.boundary) {
    this.emit('error', new Error('no suitable boundary found'))
    return this.end()
  }
  
  this.parsedLength = 0
  this.flags = 0
  this.index = null
  this.markers = {}
  this.part = null
  this.activeParts = []
  this.endEmitted = false
  this.headerField = ''
  this.headerValue = ''
  this.state = S.START
  
  request.on('error', function (err) {
    var parts = self.activeParts.slice()
    for (var p in parts) {
      var part = parts[p]
      part.end()
    }
    this.emit('error', err)
    this.end()
  })
  
  // limiting the chunkSize keeps memory usage nice and tight
  request.pipe(this, { chunkSize: this._writableState.highWaterMark })
}
util.inherits(Multiparser, stream.Writable)

Multiparser.prototype.parseHeaders = function (headers) {
  // length
  if (headers['content-length']) {
    this.length = parseInt(headers['content-length'])
  }
  
  // type & boundary
  if (headers['content-type'].match(/multipart/i)) {
    this.type = 'multipart'
    var boundary = headers['content-type'].match(/boundary=(?:"([^"]+)"|([^;]+))/i)
    if (boundary) {
      boundary = boundary[1] || boundary[2]
      this.boundary = new Buffer(boundary.length+4);
      this.boundary.write('\r\n--', 'ascii', 0);
      this.boundary.write(boundary, 'ascii', 4);
      this.lookbehind = new Buffer(this.boundary.length + 8)
      this.boundaryChars = {}
      for (var i=0; i<this.boundary.length; i++) {
        this.boundaryChars[this.boundary[i]] = true
      }
    }
  }
}

Multiparser.prototype._write = function (buffer, cb) {
  this.parsedLength += buffer.length
  this.emit('progress', this.parsedLength, this.length)
  
  var self = this,
      prevIndex = this.index,
      index = this.index,
      state = this.state,
      flags = this.flags,
      boundary = this.boundary,
      lookbehind = this.lookbehind,
      boundaryChars = this.boundaryChars,
      boundaryLength = this.boundary.length,
      boundaryEnd = boundaryLength - 1,
      bufferLength = buffer.length,
      markers = this.markers,
      i = 0,
      cl,
      c
  
  function dataCallback (name, clear) {
    if (markers[name] === undefined) return
    if (!clear) {
      callback(name, buffer, markers[name], bufferLength)
      markers[name] = 0
    } else {
      callback(name, buffer, markers[name], i)
      delete markers[name]
    }
  }

  function callback (name, buffer, start, end) {
    var name = 'on' + name.slice(0, 1).toUpperCase() + name.slice(1)
    if (start !== undefined && start === end) return
    if (self[name]) {
      self[name](buffer, start, end, function (written) {
        if (cb && !written) {
          var wcb = cb
          cb = null
          self.part.once('drain', wcb)
        }
      })
    }
  }
  
  for (i=0; i<bufferLength; i++) {
    c = buffer[i]
    switch (state) {
      case S.PARSER_UNINITIALIZED:
        return i
      case S.START:
        // skip leading CR/LF
        if (c == CR || c == LF) {
          continue
        }
        index = 0
        state = S.START_BOUNDARY
      case S.START_BOUNDARY:
        if (index == boundary.length - 2) {
          if (c != CR) {
            return i
          }
          index++
          break
        } else if (index - 1 == boundary.length - 2) {
          if (c != LF) {
            return i
          }
          index = 0
          callback('partBegin')
          state = S.HEADER_FIELD_START
          break
        }
        if (c != boundary[index+2]) {
          return i
        }
        index++
        break
      case S.HEADER_FIELD_START:
        state = S.HEADER_FIELD
        markers['headerField'] = i
        index = 0
      case S.HEADER_FIELD:
        if (c == CR) {
          delete markers['headerField']
          state = S.HEADERS_ALMOST_DONE
          break
        }
        index++
        if (c == HYPHEN) {
          break
        }
        if (c == COLON) {
          if (index == 1) {
            // empty header field
            return i
          }
          dataCallback('headerField', true)
          state = S.HEADER_VALUE_START
          break
        }
        cl = lower(c)
        if (cl < A || cl > Z) {
          return i
        }
        break
      case S.HEADER_VALUE_START:
        if (c == SPACE) {
          break
        }
        this.markers['headerValue'] = i
        state = S.HEADER_VALUE
      case S.HEADER_VALUE:
        if (c == CR) {
          dataCallback('headerValue', true)
          callback('headerEnd')
          state = S.HEADER_VALUE_ALMOST_DONE
        }
        break
      case S.HEADER_VALUE_ALMOST_DONE:
        if (c != LF) {
          return i
        }
        state = S.HEADER_FIELD_START
        break
      case S.HEADERS_ALMOST_DONE:
        if (c != LF) {
          return i
        }
        callback('headersEnd')
        state = S.PART_DATA_START
        break
      case S.PART_DATA_START:
        state = S.PART_DATA
        this.markers['partData'] = i
      case S.PART_DATA:
        prevIndex = index
        if (index == 0) {
          // boyer-moore derrived algorithm to safely skip non-boundary data
          i += boundaryEnd
          while (i < bufferLength && !(buffer[i] in boundaryChars)) {
            i += boundaryLength
          }
          i -= boundaryEnd
          c = buffer[i]
        }
        if (index < boundary.length) {
          if (boundary[index] == c) {
            if (index == 0) {
              dataCallback('partData', true)
            }
            index++
          } else {
            index = 0
          }
        } else if (index == boundary.length) {
          index++
          if (c == CR) {
            // CR = part boundary
            flags |= F.PART_BOUNDARY
          } else if (c == HYPHEN) {
            // HYPHEN = end boundary
            flags |= F.LAST_BOUNDARY
          } else {
            index = 0
          }
        } else if (index - 1 == boundary.length)  {
          if (flags & F.PART_BOUNDARY) {
            index = 0
            if (c == LF) {
              // unset the PART_BOUNDARY flag
              flags &= ~F.PART_BOUNDARY
              callback('partEnd')
              callback('partBegin')
              state = S.HEADER_FIELD_START
              break
            }
          } else if (flags & F.LAST_BOUNDARY) {
            if (c == HYPHEN) {
              callback('partEnd')
              callback('end')
              state = S.END
            } else {
              index = 0
            }
          } else {
            index = 0
          }
        }
        if (index > 0) {
          // when matching a possible boundary, keep a lookbehind reference
          // in case it turns out to be a false lead
          lookbehind[index-1] = c
        } else if (prevIndex > 0) {
          // if our boundary turned out to be rubbish, the captured lookbehind
          // belongs to partData
          callback('partData', lookbehind, 0, prevIndex)
          prevIndex = 0
          this.markers['partData'] = i
          // reconsider the current character even so it interrupted the sequence
          // it could be the beginning of a new sequence
          i--
        }
        break
      case S.END:
        break
      default:
        return i
    }
  }
  
  dataCallback('headerField');
  dataCallback('headerValue');
  dataCallback('partData');
  
  this.state = state
  this.index = index
  this.flags = flags
  
  cb && cb()
}

Multiparser.prototype.onPartBegin = function () {
  var self = this
  var part = this.part = new Part()
  part.on('end', function () {
    for (var p in self.activeParts) {
      if (self.activeParts[p] === part) {
        self.activeParts.splice(p, 1)
        break;
      }
    }
    if (self.ended && !self.endEmitted) {
      self.end()
    }
  })
  this.activeParts.push(part)
}

Multiparser.prototype.onPartData = function (b, start, end, cb) {
  var written = this.part.write(b.slice(start, end))
  cb(written)
}

Multiparser.prototype.onPartEnd = function () {
  this.part.end()
}

Multiparser.prototype.onHeaderField = function (b, start, end) {
  this.headerField += b.toString(this.encoding, start, end)
}

Multiparser.prototype.onHeaderValue = function (b, start, end) {
  this.headerValue += b.toString(this.encoding, start, end)
}

Multiparser.prototype.onHeaderEnd = function () {
  var headerField = this.headerField
  var headerValue = this.headerValue
  var part = this.part
  var m = null
  headerField = headerField.toLowerCase()
  part.headers[headerField] = headerValue
  if (headerField == 'content-disposition') {
    if (m = headerValue.match(/name="([^"]+)"/i)) {
      part.name = m[1]
    }
    m = headerValue.match(/filename="(.*?)"($| )/i)
    if (!m) return
    var filename = m[1].substr(m[1].lastIndexOf('\\') + 1)
    filename = filename.replace(/%22/g, '"')
    part.filename = filename.replace(/&#([\d]{4})/g, function (m, code) {
      return String.fromCharCode(code)
    })
  } else if (headerField == 'content-type') {
    part.mime = headerValue
  } else if (headerField == 'content-transfer-encoding') {
    part.transferEncoding = headerValue
  }
}

Multiparser.prototype.onHeadersEnd = function () {
  var part = this.part  
  this.headerField = ''
  this.headerValue = ''
  this.emit('part', part)
}

Multiparser.prototype.onEnd = function () {
  this.ended = true
  if (this.activeParts.length === 0) {
    this.endEmitted = true
    this.emit('end')
  }
}

Multiparser.prototype.end = function () {
  this.onEnd()
}


lib/transformer.js
/*
 *  transformer.js
 *
 */

// export Transformer class
module.exports = Transformer

var util = require('util')
var stream = require('stream')

function Transformer (opts) {
  stream.Duplex.call(this, opts)
  this._buffer = []
  this.once('finish', this.done.bind(this))
}
util.inherits(Transformer, stream.Duplex)

Transformer.prototype._write = function (chunk, cb) {
  var self = this
  var rs = this._readableState
  function push (err, chunk) {
    var ret = self.push(chunk)
    if (rs.length && rs.needReadable) {
      self.emit('readable')
    }
    if (!ret) {
      self._buffer.push(cb)
    } else {
      cb()
    }
  }
  if (this._transform) {
    this._transform(chunk, null, push)
  } else {
    push(null, chunk)
  }
}

Transformer.prototype._read = function (n, cb) {
  while (this._buffer.length) {
    this._buffer.shift()()
  }
}

Transformer.prototype.done = function () {
  var rs = this._readableState
  rs.ended = true
  if (rs.length && rs.needReadable) {
    this.emit('readable')
  } else if (rs.length === 0) {
    this.emit('end')
  } else {
    this.emit('drain')
  }
}


example/http.js
#!/usr/bin/env node

/*
 *  http.js
 *
 */


var fs = require('fs')
var Multiparser = require('../')
var http = require("http")
var updir = '/tmp/'

http.createServer(function (req, res) {
  if (req.url === '/upload' && req.method === 'POST') {

    var error = null
    var progress = 0
    var parser = new Multiparser(req)

    parser.on('error', function (err) {
      error = err
    })
    
    parser.on('progress', function (parsed, total) {
      var currentProgress = ~~(parsed / total * 100)
      if (currentProgress != progress) {
        console.log('uploading...', currentProgress,
                    'memory usage:', ~~(process.memoryUsage().rss / (1024 * 1024)) + 'mb')
        progress = currentProgress
      }
    })

    parser.on('part', function (part) {
      part.on('end', function () {
        if (part.filename) {
          console.log('uploaded file "' + part.name + '" to ' + updir + part.filename)
        } else if (part.value) {
          console.log('parsed field "' + part.name + '" as "' + part.value + '"')
        }
      })
      
      if (part.filename) {
        var file = fs.createWriteStream(updir + part.filename)
        part.pipe(file)
      } else {
        part.value = ''
        part.on('readable', function () {
          part.value += part.read()
        })
        part.read(0)
      }
    })

    parser.on('end', function () {
      if (error) {
        res.writeHead(500)
        return res.end(error.message)
      } else {
        res.end('all parsed!')
      }
    })
    
  } else {
    
    res.end('<html>\
      <form enctype="multipart/form-data" method="POST" action="/upload">\
      <input name="fieldup" type="text" /><br>\
      <input name="fileup" type="file" multiple="true" /><br>\
      <input type="submit" value="upload" />\
      </form>\
      </html>')
      
  }
}).listen(8080)
